File: concurrent_kata/mrugger/ruby/test.stream.rb

class TestStream

  @sequence

  @total_sequence_length
  @current_position

  def initialize(sequence, replication_count)
    @sequence = sequence
    @total_sequence_length = sequence.length * replication_count
    @current_position = 0
  end

  def readbytes(count, block)
    copy_count = 0
    copy_count += copy_next_block(count - copy_count, block) while (copy_count < count && !eof?)
  end

  def rewind(count)
    @current_position = (@current_position > count) ? @current_position - count : 0
  end

  def eof?
    @current_position >= @total_sequence_length
  end


  private

  def copy_next_block(count, block)
    begin_index = @current_position % @sequence.length

    # Which ever is smaller
    copysize = [count,                                        # the requested number of characters
                @total_sequence_length - @current_position,   # what's left of the total replicated sequence
                @sequence.length - begin_index].min           # what's left in the sequence

    end_index = begin_index + copysize - 1

    block << @sequence[begin_index..end_index]
    @current_position += copysize
    return copysize
  end

end

File: concurrent_kata/mrugger/cplusplus/test.stream.cpp
#include "test.stream.h"


TestStream::TestStream(std::string &sequence, unsigned int replication_count)
: _sequence(sequence),
  _replication_count(replication_count),
  _current_position(0),
  _total_sequence_length(sequence.length() * replication_count)
{}


unsigned int TestStream::readbytes(unsigned int count, char *block)
{
  unsigned int copy_count = copy_next_block(count, block);

  while (copy_count < count && !is_eof())
    copy_count += copy_next_block(count, block + copy_count);

  return copy_count;
}


void TestStream::rewind(unsigned int count)
{
  _current_position = (_current_position > count) ? _current_position - count : 0;
}


bool TestStream::is_eof(void)
{
  _current_position >= _total_sequence_length;
}


unsigned int TestStream::copy_next_block(unsigned int count, char *block)
{
  unsigned int begin_index = _current_position % _sequence.length();
  std::size_t copied = _sequence.copy(block, count, begin_index);
  _current_position += (unsigned int) copied;
  return copied;
}
